// @lemonadejs/rating/dist/vue.js
import { h } from 'vue';
import component from "./index";


export const Rating = {
    inheritAttrs: false,
    mounted() {
        let options = {
            ...this.$attrs
        };

        this.el = this.$refs.container;

        this.current = component(this.$refs.container, options);
    },
    setup() {
        let containerProps = {
            ref: 'container',
            style: {
                width: '100%',
                height: '100%',
            }
        };
        return () => h('div', containerProps);
    },
    watch: {
        $attrs: {
            deep: true,
            handler() {
                this.updateState();
            }
        }
    },
    methods: {
        updateState() {
            for (let key in this.$attrs) {
                if (this.$attrs.hasOwnProperty(key) && this.current.hasOwnProperty(key)) {
                    if (this.$attrs[key] !== this.current[key]) {
                        this.current[key] = this.$attrs[key];
                    }
                }
            }
        }
    }
}

export default Rating;
